### PR TITLE
Refactored mod creation to use a file tree of templates.

### PR DESCRIPTION
**Pull request by Mirsario**
To assist two+ active PRs (#2472 and #4037), this one does the following:
- [x] Moves code related to creating & upgrading mod source folders out of UI and into a new separate `Terraria.ModLoader.Core.SourceManagement` class.
- [x] Replaces textfile-generating hardcode with a filetree of embedded resources that are enumerated and exported to populate mod src folders. The embedded resources use the [Mustache](https://mustache.github.io/mustache.5.html) logicless template format in their names and contents (this concept will be nice to use for mod descriptions too in userland in the future). Optional files, such as the BasicSword, are skipped by making their filenames evaluate to nothing. 
- [x] Removes `TargetFramework` and `PlatformTarget` from default `.csproj` files, putting it into ~~`tModLoader.targets`~~ `tMLMod.targets` instead.
- [x] Makes Csproj upgrades attempt to function via XML edits, if they're not too broken.
- [ ] When this hits Stable, update https://github.com/tModLoader/tModLoader/wiki/Basic-tModLoader-Modding-Guide#mod-skeleton-contents info and screenshots with new `Content.Items` namespace and folder.

### Default mod template changes:
- `launchSettings.json`, uses tabs now.
- Csproj uses tabs now, no longer has an obsolete <?xml> header, is now practically empty besides for the targets import, with every other line just being nice "put this here" markups.
- The item template now uses the `ExampleSword.png`, which is nicer to look at. A tooltip has also been added, to immediately showcase how to make multiline strings in HJSON.
- The two `.cs` files now use file-scoped namespaces. The item template's code had its comments updated, and had magic numbers replaced with ID classes in a few lines.
- `description.txt`:
```diff
-{{ModDisplayName}} is a pretty cool mod, it does...this. Modify this file with a description of your mod.
+Modify this file with a description of your mod.
```
- As too many people don't bother even providing a mod icon, I felt like updating the default mod icon to be slightly more presentable, while communicating its placeholder nature:
 -> 

### Questionable notes:
- I've had to exclude the new template folder from `AssemblyResourcesContentSource`, because of it rather annoyingly crashing the game whenever it encountered a `.cs` + `.png` pair.
- The templates' EmbeddedResource LogicalNames are set to use forward slashes, due to localization files utilizing dots, which slashes get reasonlessly converted to by default. Ideally even the vanilla game would convert to using forward slashes, it's beneficial for its asset system.

### CODE REVIEW

**Comment by Destructor-Ben** [comment]
In a future PR could we allow specifying our own template? atm I copy paste a template and hope I remembered to change all file names + contents of files

**Comment by JavidPack** [comment]
Should we move LangVersion to the tMLMod.targets too? I believe a modder should still be able to override it in their mod's csproj, but an issue we have right now is that modders are seeing suggestions in VS for language versions that the ingame complier does not support. I know there is a pr for doing it through dotnet itself, but even then we'd want to make sure that only the highest supported lang version is the default.

**Comment by Mirsario** [comment]
> Should we move LangVersion to the tMLMod.targets too?

I've done that now, with it explicitly set to `10.0` for now. But only overrides with old versions in values will be auto-removed in the upgrade process. `latest`, `preview`, and >=`11.0` will be kept, so I guess it's not really helping what you've described. Oh well, hopefully we merge 2472 soon.

**Comment by JavidPack** [comment]
Will this be able to easily remove `<PackageReference Include="tModLoader.CodeAssist" Version="0.1.*" />` once #4099 is merged? I haven't parsed all the code yet, just curious.

**Comment by Mirsario** [comment]
Yes, here's the code to insert even:
```cs
// Remove the analyzer package, since our targets file now handles all that.
RemoveNodes(itemGroups.Elements("PackageReference").Where(
	e => e.Attribute("Include")?.Value == "tModLoader.CodeAssist"
));
```

**EDIT:** This is now a part of the PR, with the PackageReference moved into `tMLMod.targets`.

File: patches/tModLoader/Terraria/ModLoader/Assets/AssemblyResourcesContentSource.cs
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/ModLoader/Assets/AssemblyResourcesContentSource.cs:29]: Is this necessary? You're not using the asset system for your templates anymore

File: patches/tModLoader/Terraria/ModLoader/Templates/Properties/launchSettings.json
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/ModLoader/Templates/Properties/launchSettings.json]: Why would the average user want these two configurations?

File: patches/tModLoader/Terraria/ModLoader/Templates/Properties/launchSettings.json
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/ModLoader/Templates/Properties/launchSettings.json]: Don't these need to use `$(DotNetName)`?

File: patches/tModLoader/Terraria/ModLoader/Templates/{{ModName}}.csproj
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/ModLoader/Templates/{{ModName}}.csproj]: Remove?

File: patches/tModLoader/Terraria/ModLoader/Templates/{{ModName}}.csproj
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/ModLoader/Templates/{{ModName}}.csproj]: Does this not belong in `tMLMod.targets` too?

File: patches/tModLoader/Terraria/Terraria.csproj.patch
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/Terraria.csproj.patch:86]: You can use `Update=` instead of `Remove` and then `Include`

File: patches/tModLoader/Terraria/ModLoader/Templates/{{ModName}}.cs
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/ModLoader/Templates/{{ModName}}.cs]: Needs approval from @JavidPack are we doing file scoped namespaces now? We don't in ExampleMod

File: patches/tModLoader/Terraria/ModLoader/Core/SourceManagement.cs
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/ModLoader/Core/SourceManagement.cs]: why use record struct and boxing, rather than regular `record`

File: patches/tModLoader/Terraria/ModLoader/Core/SourceManagement.cs
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/ModLoader/Core/SourceManagement.cs]: This needs to be a modification itself. 
I also prefer `modifications.Any()`

File: patches/tModLoader/Terraria/ModLoader/Core/SourceManagement.cs
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/ModLoader/Core/SourceManagement.cs]: this is currently the only way to get `FileBroken`. See my above comment about factoring this out, so this method doesn't need a `bool` return, and can just return (or `yield`) the list of modifications instead

File: patches/tModLoader/Terraria/ModLoader/Core/SourceManagement.cs
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/ModLoader/Core/SourceManagement.cs]: I feel like the `status` enum here is not particularly needed or readable. 

```cs
if (File.Exists(csprojPath)
      && TryLoadXmlDocument(csprojPath, LoadOptions.PreserveWhitespace, out var document)
      && IsModernCsproj(document)) {
    
    var modifications = CollectCsprojUpgradeActions(document);
    if (modifications.Count == 0)
        return false;

    action = () => {
        modifications!.ForEach(m => m());
        WriteXmlDocumentToFile(csprojPath, document!);
    }
    return true;
}

action = () => ResetCsprojFile(csprojPath, templateParameters);
return true;
```

File: patches/tModLoader/Terraria/ModLoader/Core/SourceManagement.cs
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/ModLoader/Core/SourceManagement.cs]: No need for the `createBackup` parameter, just make a backup inside the function if the file exists

File: patches/tModLoader/Terraria/ModLoader/Core/SourceManagement.cs
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/ModLoader/Core/SourceManagement.cs]: Normally I'd factor this as a local function at the top of the method, like:
```cs
void RemoveNodes(IEnumerable<XNode> nodes) {
    modifications.Add(() => ...)
}
```

**Review by Chicken-Bones — CHANGES_REQUESTED** [review]
Overall this is an excellent change. I've provided some factoring feedback.

This PR should _not_ change the templates. It makes it really hard to review. Please split template changes into an additional 1 or 2 PRs (2 if there are changes that have to be done both before and after this PR).

File: patches/tModLoader/Terraria/ModLoader/Assets/AssemblyResourcesContentSource.cs
Comment (Mirsario) [reply patches/tModLoader/Terraria/ModLoader/Assets/AssemblyResourcesContentSource.cs:29]: Duplicate extension errors are thrown at ContentSource initialization time, not access time. Any duplicate extensions found in EmbeddedResources will crash the game. This behavior seriously needs redesigns, as we still have crashes from some resourcepacks intended for vanilla.

File: patches/tModLoader/Terraria/ModLoader/Templates/Properties/launchSettings.json
Comment (Mirsario) [reply patches/tModLoader/Terraria/ModLoader/Templates/Properties/launchSettings.json]: I guess it'll cause more confusion than it's worth, will remove.

File: patches/tModLoader/Terraria/ModLoader/Core/SourceManagement.cs
Comment (Mirsario) [reply patches/tModLoader/Terraria/ModLoader/Core/SourceManagement.cs]: A CA rule cries about .Any() being inferior. csharpbad.

**Pull request merged**
